### PR TITLE
CircleCI pushes releases to dockerhub based on tags and branches

### PR DESCRIPTION
**Pull request by dimroc**
https://www.pivotaltracker.com/n/projects/2235105/stories/168396064

image

Key features are the support for git tags.

1. The following builds and releases a version on master.

Chainlink Core only:
```
git tag v0.0.2
git push origin v0.0.2
```

Explorer only:
```
git tag explorer-v0.0.2
git push origin explorer-v0.0.2
```

2. Develop branch will update images with the tag `developer`

3. Release candidates will be built when pushing to a branch like `release/0.0.2`, resulting in an image `smartcontract/chainlink:candidate-0.0.2` and `smartcontract/explorer:candidate-0.0.2`.

4. What isn't in this final PR, but we could readd, is an escape hatch for testing and debugging. I allowed `chores-v.*` to be built via tags in order for me to have confidence in the PR. If @thodges-gh you see a need for something like this, let me know.

### CODE REVIEW

**Comment by dimroc** [comment]
Successfully builds on tags

image

File: tools/bin/circleci_push_explorer [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+set -ex
```
Comment (dimroc) [inline tools/bin/circleci_push_explorer]: Deliberately left as `set -ex` to have insight on CI.

File: tools/bin/circleci_push_chainlink [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# to work with.
+#
+
+circle_branch=$1
+circle_tag=$2
+
+PROJECT_ROOT=`git rev-parse --show-toplevel`
+pushd $PROJECT_ROOT >/dev/null
+
+tools/bin/circleci_validate_if_versioned $circle_branch $circle_tag
```
Comment (felder-cl) [inline tools/bin/circleci_push_chainlink]: Maybe surround the bash environment variables with "quotation marks"?

File: tools/bin/circleci_validate_if_versioned [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+if [ "$circle_branch" == "master" ] && [ -z "$circle_tag" ]
+then
+  echo On master with no git tag version, erroring.
+  exit 1
+fi
+
+# if version tagged, but not master, error. Only master can have versions.
+if [ "$circle_branch" != "master" ] && [ -n "$circle_tag" ]
+then
+  echo Not on master \(${circle_branch}\) with git tag version \(${circle_tag}\), erroring.
```
Comment (felder-cl) [inline tools/bin/circleci_validate_if_versioned]: maybe quotations?

File: tools/bin/gittag2dockertag [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+set -e
+
+if [ -z "$1" ]
+then
+  exit 0 # No tag, skip.
+fi
```
Comment (felder-cl) [inline tools/bin/gittag2dockertag]: if [ -z "$1" ]; then ?

File: .circleci/config.yml [hunk 5]
```diff
@@ -236,14 +234,22 @@ jobs:
...
+            echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USER" --password-stdin
       - run:
           name: Docker build
-          command: docker build -f explorer/Dockerfile .
-  build-and-publish-chainlink:
+          command: docker build -f explorer/Dockerfile -t smartcontract/explorer:circleci .
+      - run:
+          name: Docker push, if applicable
+          command: |
+            tools/bin/circleci_push_explorer ${CIRCLE_BRANCH} ${CIRCLE_TAG}
```
Comment (j16r) [inline .circleci/config.yml]: Should this be in the `ci` directory we have in tools/?

File: tools/bin/branch2tag [hunk 6]
```diff
@@ -7,10 +7,8 @@ if [ "${branch}" = "develop" ]; then
   echo develop
 elif [ "${branch}" = "master" ]; then
   echo latest
-elif [[ "${branch}" =~ ^hotfix\/([a-zA-Z0-9.]+) ]]; then
-  echo hotfix-${BASH_REMATCH[1]}
 elif [[ "${branch}" =~ ^release\/([a-zA-Z0-9.]+) ]]; then
-  echo "${BASH_REMATCH[1]}"
+  echo "candidate-${BASH_REMATCH[1]}"
```
Comment (j16r) [inline tools/bin/branch2tag]: 👍

**Comment by felder-cl** [comment]
looks good as far as I can tell 👍

**Pull request merged**